hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -3394,6 +3396,42 @@ private CopyResult copyFile(String srcKey, String dstKey, long size,
         });
   }
 
+  /**
+   * Propagate encryption parameters from source file if set else use the
+   * current file system encryption settings.
+   * @param srcom
+   * @param copyObjectRequest
+   */
+  private void propagateEncryptionParams(ObjectMetadata srcom,
+                                         CopyObjectRequest copyObjectRequest) {
+    Optional<SSEAwsKeyManagementParams> kmsParams = Optional.empty();
+    String sourceKMSId = srcom.getSSEAwsKmsKeyId();
+    if (isNotEmpty(sourceKMSId)) {
+      // source KMS ID is propagated
+      LOG.debug("Propagating SSE-KMS settings from source {}",
+          sourceKMSId);
+      kmsParams = Optional.of(new SSEAwsKeyManagementParams(sourceKMSId));
+    }
+    kmsParams.ifPresent(
```

Comment:
> pull that up into the if() clause and you can avoid doing the optional work, just
> setSSE...(new SS3AwsKMP(sourceKMSId)).
Done
> now, troublespot, and its one I'm curious about. What if there's SSE-C set, as it is also being set on the request? FWIW, I think things will break trying to read the file by setting the SSE-C key will inevitably break too.
If SSE-C is used, sseKmsKey won't be present in the sourceObjectMeta. So that shouldn't be a problem. I debugged this test ITestS3AEncryptionSSEC#testRenameFile and it works fine. Please let me know if I am missing something.